src/core/common/instance.hpp

```diff
@@ -395,7 +418,10 @@ class Instance : public otInstance, private NonCopyable
 #if OPENTHREAD_CONFIG_DIAG_ENABLE
     FactoryDiags::Diags mDiags;
 #endif
-    bool mIsInitialized;
+    bool mIsInitialized : 1;
+#if OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE && (OPENTHREAD_FTD || OPENTHREAD_MTD)
+    bool mUncompressedDnsName : 1;
```

Comment:
I guess this does not work in `MULTIPLE_INSTANCE` because we can't call `Instance::Get().IsUncompressedDnsNameEnabled()` in `MULTIPLE_INSTANCE`. 

I am fine with this feature not working in MULTIPLE_INSTANCE, but can we just make this field static so that we don't need to explicitly specify that this feature does not work in `MULTIPLE_INSTANCE`.  

```suggestion
    static bool mUncompressedDnsName : 1;
```

Reply:
I was also considering the same (since defining `mUncompressedDnsName` as `static` member variable fits well here) but decided not to go for it.
 
I think unless we have a very good reason it would be good to avoid using `static` member variable in any of OT classes (and mainly instance) so that `sizeof(ot::Instance)` can give us  the OT RAM/memory would use.

Reply:
> `sizeof(ot::Instance)` can give us the OT RAM/memory would use.

Any elaboration on why this is useful?

Reply:
In general, in terms of sw architecture, it is good practise to have all variables (memory use) for OT library to follow the same pattern.

One use-case is to be able to easily tell how much RAM OT lib is using when OT is integrated in a production sw (with a selected set of production/project configs). The `static` sort of hides the memory use by the OT library.

We have this related point in our style guide:

> - Global and Scoped Static Construction
>   - There shall be no use of global, static or otherwise, object construction. The use of scoped static object construction should be avoided.
>

Reply:
I would agree on the principle. However, we don't turn on `REFERENCE_DEVICE` for production sw. So, defining static variables for `REFERENCE_DEVICE` is actually fine?

Reply:
It sets a precedence for use of `static` and I think we should have a good justification/advantage when we do so.

In the case here the main advantage is that we we want to limit the definition and more importantly the  use of the related API under `MULTIPLE_INSTANCE`. We can already do so by adding proper `#if` checks (which we do in the `dns_api.cpp` in this PR).

Reply:
> It sets a precedence for use of static and I think we should have a good justification/advantage when we do so.

We already did that:
```c++
#if !OPENTHREAD_CONFIG_HEAP_EXTERNAL_ENABLE
    static Utils::Heap sHeap;
#endif
```

> In the case here the main advantage is that we we want to limit the definition and more importantly the use of the related API under `MULTIPLE_INSTANCE`. 

Limiting the definition and use of related APIs under `MULTIPLE_INSTANCE` seems unnecessary to me. If less code can just work for `MULTIPLE_INSTANCE` and non `MULTIPLE_INSTANCE`, why should we add limitations with more code?

---

That being said. It's a minor issue for me. But just prefer to make it also usable in `MULTIPLE_INSTANCE` when there is no obvious reason why not.

Reply:
I changed this in last push.
Somehow I did not notice that with static we can get it to work under `MULTIPLE_INSTANCE` too. I think that may be good enough reason to allow it here.